Add an option for setup:config:set to define the document root

AntonEvers opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Add `bin/magento setup:config:set --document-root-is-pub`
I think this is needed in order to raise awareness for https://devdocs.magento.com/guides/v2.3/install-gde/tutorials/change-docroot-to-pub.html section 3, and to simplify automating this setting in deployment scripts.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#29523: Add an option for setup:config:set to define the document root

AntonEvers committed:
Add an option for setup:config:set to define the document root

Add unit tests and allow to set true or false

clean up interactive options and complet unit test

set document-root-is-pub to false by default so that it is always
present after installation in the env.php file

resolve static tests

process code review comments

remove duplicate word root

add boolean selector for interactive installations

for backward compatibility, allow an empty value on booleans

assign boolean config to document-root-is-pub

resolve: PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Magento\Framework\Setup\Option\BooleanConfigOption::boolVal() must be of the type string, int given

rename Directories to Directory

resolve test issues

resolve Magento Health Index test results

remove static boolVal

fix unit test

resolve issue that led to major version change

simplify tests

resolve integration tests

AntonEvers commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
@AntonEvers good idea!
Could you cover this new command with any kind of test? I think unit will be ok there.

Based on the description i have few questions:
1. What if we'll run just `bin/magento setup:config:set --document-root-is-pub`? Will it set to "true" or it will ask for setting true / false?
2. What if we'll run just `bin/magento setup:config:set --document-root-is-pub true`? Will it set to "true"?
3. What if we'll run `bin/magento setup:config:set --document-root-is-pub false`? Will it set it as "false"?

Would be nice to have more clear  description for the command option

AntonEvers commented:
> @AntonEvers good idea!
> Could you cover this new command with any kind of test? I think unit will be ok there.
> 
> Based on the description i have few questions:
> 
> 1. What if we'll run just `bin/magento setup:config:set --document-root-is-pub`? Will it set to "true" or it will ask for setting true / false?
> 2. What if we'll run just `bin/magento setup:config:set --document-root-is-pub true`? Will it set to "true"?
> 3. What if we'll run `bin/magento setup:config:set --document-root-is-pub false`? Will it set it as "false"?
> 
> Would be nice to have more clear description for the command option

Hi @ihor-sviziev 
The command option currently works similar to `--allow-parallel-generation`, `--skip-db-validation`, `--db-ssl-verify`, `--cleanup-database` and `--use-sample-data`.
It is a flag, rather than an option and it does not allow a value. `bin/magento setup:config:set --document-root-is-pub` would add the following to `env.php`:
```php
'directories' => [
    'document_root_is_pub' => true
]
```
I will add some unit testing. Please let me know if the command should behave more like `--use-secure=USE-SECURE` instead, to allow setting it both to `true` or `false`.

AntonEvers commented:
I changed the command to a boolean select option.
During interactive installation it will now asK:
```
Is the webserver's document root set to ./pub/?
  [0] no
  [1] yes
```
It accepts inputs: yes, no, 1, 0, true, false.

AntonEvers commented:
Now I have set the default value to false, so that the key is always present in env.php. This should make developers more aware of the fact that this option exists.

ihor-sviziev commented:
From first look - looks good!
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @AntonEvers ,
Could you fix static tests? Everything else looks good

AntonEvers commented:
@magento run Static Tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * @inheritdoc
+     */
+    public function validate(array $options, DeploymentConfig $deploymentConfig)
```
Until it get merged - can we add there return types?
```suggestion
    public function validate(array $options, DeploymentConfig $deploymentConfig): array
```

ihor-sviziev commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * @inheritdoc
+     */
+    public function getOptions()
```
Can we define here return type?
```suggestion
    public function getOptions(): array
```

ihor-sviziev commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * @inheritdoc
+     */
+    public function createConfig(array $options, DeploymentConfig $deploymentConfig)
```
Can we define here return type?

ihor-sviziev commented on setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * testGetOptions
+     */
+    public function testGetOptions()
```
```suggestion
    public function testGetOptions(): void
```

ihor-sviziev commented on setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * testCreateConfigCacheRedis
+     */
+    public function testCreateConfig()
```
```suggestion
    public function testCreateConfig(): void
```

ihor-sviziev commented on setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * testCreateConfigCacheRedis
+     */
+    public function testCreateConfigInvalid()
```
```suggestion
    public function testCreateConfigInvalid(): void
```

ihor-sviziev commented on setup/src/Magento/Setup/Test/Unit/Model/ConfigOptionsList/DirectoriesTest.php:
```diff
@@ -0,0 +1,93 @@
+     */
+    public function testCreateConfig()
+    {
+        $tests = [
```
Could you move it to data provider? Could you move it to data provider?

ihor-sviziev commented:
Looks like that’s last part of requests changes, sorry for not requesting them earlier.

AntonEvers commented:
No problem at all @ihor-sviziev!

AntonEvers commented:
@magento run all tests

ihor-sviziev reviewed (dismissed):
✔ Approved.

Failing tests looks not related to changes form this PR.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

engcom-Alfa commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Directories.php:
```diff
@@ -0,0 +1,85 @@
+use Magento\Framework\Setup\Option\SelectConfigOption;
+
+/**
+ * Deployment configuration options for the document root root
```
@AntonEvers Could you delete the duplicate word - **root**? Thanks!

engcom-Alfa reviewed (changes requested):
Hi, @AntonEvers . During testing, we faced some issues.

**Case 1.** Unable to set true/false configuration value with **0** and **1**

Run in console: `bin/magento setup:config:set --document-root-is-pub 1` for example;

**Actual Result:** :heavy_multiplication_x: Isn't supported value '1'

![Screenshot from 2020-07-28 14-58-52](https://user-images.githubusercontent.com/51679138/88662639-e74d2380-d0e2-11ea-97b2-146ec396d3ae.png)

**Case 2.** Unable to set true/false configuration without value

Run in console: `bin/magento setup:config:set --document-root-is-pub` 

**Actual Result:** :heavy_multiplication_x: The option requires a value

![Screenshot from 2020-07-28 14-26-53](https://user-images.githubusercontent.com/51679138/88662954-6e020080-d0e3-11ea-9872-e8ed681a3554.png)

Although, if we run the command like `bin/magento setup:config:set --allow-parallel-generation` for example, :heavy_check_mark:  we can choose **Y/n**

![Screenshot from 2020-07-28 15-03-57](https://user-images.githubusercontent.com/51679138/88663203-df41b380-d0e3-11ea-8eb5-f20fe244df72.png)

@AntonEvers Could you take a look?

Thanks!

AntonEvers commented:
@engcom-Alfa I would like to connect this PR with https://github.com/magento/magento2/pull/29187 in order to fix the 0/1 issue. Can you please help me figure out what is going on with the tests there?

AntonEvers commented:
@magento combine 29187 29176

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@ihor-sviziev @engcom-Alfa something in the addition of a BooleanConfigOption is messing up all of the tests, most likely during the setup commands that are part of the test, but I have no insight in that. Can you help me figure this one out?

ihor-sviziev commented:
@magento combine 29187

magento-engcom-team commented:
Hi @ihor-sviziev. Thank you for your request. I'm working on combining the pull requests for you

magento-engcom-team commented:
@AntonEvers all pull requests have been successfully combined together: 
 - merged https://github.com/magento/magento2/pull/29187: add boolean selector for interactive installations

ihor-sviziev commented:
@slavvka @lenaorobei could you add some details why tests fails? Seems like we have some fatal error?

lenaorobei commented:
Yes, there is an error during installation.

```
[2020-07-31 09:23:26] report.ERROR: There are no commands defined in the "deploy:mode" namespace.
#0 /var/www/html/vendor/symfony/console/Application.php(650): Symfony\Component\Console\Application->findNamespace()
#1 /var/www/html/vendor/symfony/console/Application.php(235): Symfony\Component\Console\Application->find()
#2 /var/www/html/lib/internal/Magento/Framework/Console/Cli.php(115): Symfony\Component\Console\Application->doRun()
#3 /var/www/html/vendor/symfony/console/Application.php(147): Magento\Framework\Console\Cli->doRun()
#4 /var/www/html/bin/magento(23): Symfony\Component\Console\Application->run()
#5 {main} [] []
```

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@magento run Static Tests, Unit Tests, Integration Tests, Functional Tests B2B

AntonEvers commented:
@magento run Integration Tests

AntonEvers committed:
Merge branch '2.4-develop' into document_root_is_pub

AntonEvers commented:
@magento run all tests

AntonEvers commented:
@ihor-sviziev Static and Functional test fails are not caused by this PR.

ihor-sviziev committed:
Fix static test

ihor-sviziev commented:
@magento run Static Tests

ihor-sviziev committed:
Fix static test

ihor-sviziev commented:
@magento run Static Tests

AntonEvers committed:
Fix static test

AntonEvers commented:
@magento run Static Tests

ihor-sviziev commented:
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

engcom-Alfa reviewed (changes requested):
Hi @AntonEvers . During testing, we faced some issues.

**Case 1.** Unable to set value for `--consumers-wait-for-messages`

Run in console: `bin/magento setup:config:set --consumers-wait-for-messages=0` (or **no** or **false**)

**Actual Result:** :heavy_multiplication_x: The value does not change

![2020-08-10_13-39](https://user-images.githubusercontent.com/51679138/89774629-e965a800-db0e-11ea-9b4b-2bdc9268f5b7.png)

The same situation with enabling option.

**Case 2.** The value for the `--allow-parallel-generation` option is stored in the wrong format

Run in console: `bin/magento setup:config:set --allow-parallel-generation 1` or **no** or **false**

**Actual Result:** :heavy_multiplication_x: All values are stored in `' '`

![2020-08-10_13-31](https://user-images.githubusercontent.com/51679138/89775081-c091e280-db0f-11ea-86b3-72a4538e34a7.png)
![2020-08-10_13-32](https://user-images.githubusercontent.com/51679138/89775084-c38cd300-db0f-11ea-98e8-4dbdef9fac33.png)
![2020-08-10_13-32_1](https://user-images.githubusercontent.com/51679138/89775093-ca1b4a80-db0f-11ea-8f35-ab603179ba21.png)

According to the Magento [documentation](https://devdocs.magento.com/guides/v2.4/config-guide/redis/redis-pg-cache.html) should be as follows:
![Screenshot from 2020-08-10 13-47-39](https://user-images.githubusercontent.com/51679138/89775237-149cc700-db10-11ea-91b1-c5eb82661843.png)

@AntonEvers Could you take a look?

Thanks!

AntonEvers committed:
resolve QA defect

AntonEvers commented:
@magento run all tests

AntonEvers committed:
fix unit test

AntonEvers commented:
@magento run all tests

AntonEvers committed:
resolve phpstan error

AntonEvers commented:
@magento run all tests

sidolov commented:
@magento create issue

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

ihor-sviziev commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Cache.php:
```diff
@@ -293,6 +299,19 @@ private function getDefaultConfigValue($inputKey)
         return substr(\hash('sha256', dirname(__DIR__, 6)), 0, 3) . '_';
     }
+
+    /**
+     * Convert any valid input option to a boolean
+     *
+     * @param mixed $option
+     *
+     * @return int|null
+     */
+    private function boolVal($option): ?int
+    {
+        return (int)in_array(strtolower((string)$option), BooleanConfigOption::OPTIONS_POSITIVE);
+    }
```
```suggestion
    /**
     * Convert any valid input option to a boolean
     *
     * @param mixed $option
     *
     * @return bool
     */
    private function boolVal($option): bool
    {
        return in_array(strtolower((string)$option), BooleanConfigOption::OPTIONS_POSITIVE);
    }
```

Accodring to https://github.com/magento/magento2/blob/5daadaee54d523c9f6bb3939bcb2df70539d8795/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php#L121-L127 and 
https://github.com/magento/magento2/blob/5daadaee54d523c9f6bb3939bcb2df70539d8795/lib/internal/Magento/Framework/Cache/LockGuardedCacheLoader.php#L142-L144

this value should be always boolean

ihor-sviziev commented on setup/src/Magento/Setup/Model/ConfigOptionsList/Directory.php:
```diff
@@ -8,83 +8,71 @@
+            if (!isset($options[$inputKey])) {
+                continue;
+            }
+            $configData->set($configPath, (int)$this->boolVal($options[$inputKey]));
```
I'm not sure how unit test not failing, but seems like it should be like that:
```suggestion
            $configData->set($configPath, $this->boolVal($options[$inputKey]));
```

Could you double check the test?

ihor-sviziev reviewed (changes requested):
@AntonEvers 
I got feedback from @engcom-Alfa that it still setting 1/0 for the `allow_parallel_generation` and `document_is_root`, while docs saying they have to be boolean.

Could you review my comments?
https://github.com/magento/magento2/pull/29176#pullrequestreview-471464208
https://github.com/magento/magento2/pull/29176#pullrequestreview-471464715

ihor-sviziev committed:
Add an option for setup:config:set to define the document root

Fix issue with incorrect type set for allow_parallel_generation and document_is_root params

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@AntonEvers I just fixed listed above issues. 
FYI assertSame correctly checks all types, while assertEquals things that '0' and false are the same.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7889](https://jira.corp.magento.com/browse/ENGCOM-7889) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

Run in console: `bin/magento setup:config:set --document-root-is-pub` **1** or **true** or **yes**

**After:** :heavy_check_mark: the values ​​are written in the correct format (boolean)

![Screenshot from 2020-09-10 16-22-30](https://user-images.githubusercontent.com/51679138/92735168-d8b08800-f381-11ea-979d-636550502cd6.png)

Run in console: `bin/magento setup:config:set --allow-parallel-generation` **1** or **true** or **yes**

**After:** :heavy_check_mark: the values ​​are written in the correct format (boolean)

![Screenshot from 2020-09-10 16-26-13](https://user-images.githubusercontent.com/51679138/92735792-5f656500-f382-11ea-82c7-956e8e273cd6.png)

Run in console: `in/magento setup:config:set --consumers-wait-for-messages **0** or **false** or **no**


**After:** :heavy_check_mark: the values ​​are written in the correct format (1/0)

![Screenshot from 2020-09-10 16-27-43](https://user-images.githubusercontent.com/51679138/92736032-93d92100-f382-11ea-91d1-dce3f6f92db4.png)

gabrieldagama commented:
@magento run all tests

Commit:
Merge branch '2.4-develop' into document_root_is_pub

gabrieldagama commented:
@magento run all tests

Commit:
Merge branch '2.4-develop' into document_root_is_pub

sidolov commented:
@magento run all tests

ihor-sviziev commented:
@gabrieldagama @sidolov is there any reasons why this PR isn't getting merged?

gabrieldagama commented:
@ihor-sviziev We are having some issues with its tests that are being looked at internally. Just for reference, the internal ticket is https://jira.corp.magento.com/browse/MTS-1593

ihor-sviziev commented:
Hi @AntonEvers,
According to the recent changes in 2.4-develop - the support of non-pub was removed and now Magento always uses the pub directory.
https://github.com/magento/magento2/commit/640cad53009b291334234ccd61ab79f256b43da2#diff-2323d6157444b3cf24d4d76635f27185a4e6ad979b0ec2af53e6592d3ca465d5R13

Due to this change I think we could close this PR as not relevant anymore.
 What do you thing?

hostep commented:
@ihor-sviziev: nice find!

But I really don't understand how they can make such a major change in a minor release? This change will probably be scheduled for Magento 2.4.2, even though it should be for 2.5.0. This might break a lot of installations.

I'm not saying it's a bad thing, it should have happened much earlier, but big changes like these really shouldn't be done in a minor release.

(sorry to interrupt here in this PR)

ihor-sviziev commented:
@sivaschenko @sidolov could you provide more info why this commit was merged to 2.4-develop as it's introducing backward incompatible changes? https://github.com/magento/magento2/commit/640cad53009b291334234ccd61ab79f256b43da2#diff-2323d6157444b3cf24d4d76635f27185a4e6ad979b0ec2af53e6592d3ca465d5R13

shiftedreality commented:
@hostep we updated Apache config to use pub directory. Since Apache reads configs on-fly, we didn't notice any issue having it updated.

This solves a few issues:
- Deprecates usage of `\Magento\Config\Model\Config\Reader\Source\Deployed\DocumentRoot`. This class was added but not used in the places where it should.
- Standardizes pub directory paths all over magento without class above
- Standartazise both Apache and Nginx entrypoints

Without this change, we saw a lot of issues while adding RemoteStorage functionality and saw possible issues for developers who work with FS API.

Could you provide any possible risks which we might miss?

hostep commented:
@shiftedreality: I understand why this is a good improvement, it really is, I and many others have been wanting it even [2 years before](https://github.com/magento/magento2/issues/372#issuecomment-25754683) Magento 2.0.0 was launched.

But where I'm seeing problems is that certain users now will be surprised when they perform a minor update and notice they have to change their document root and somehow can't. I'm pretty sure that there are (cheap) hosting companies around which don't allow you to change the document root and make it very difficult to host magento from inside the `pub` directory. Which is a shame and people shouldn't use these sort of hosting companies, but yes, a lot of people want cheap hosting... So I think it only makes sense to introduce such a big change in a major release like 2.5.0 to avoid breaking shops just because they made a wrong decision in choosing a certain hosting company.
If that's really not possible and it needs to be shipped in 2.4.x, then the release notes will have to add a very big warning on top (and not hidden in the hundreds of changes) so that users are aware of this change.

shiftedreality commented:
@hostep my understanding is those Customers have an Apache which reads `.htaccess` file. So, it should be supported fine, since the file is updated. Isn't this standard case for Apache on hosting companies?

hostep commented:
The `.htaccess` file in the Magento root is deleted, so I'm not sure what you are saying?
Also, you can not change the document root from a `.htaccess` file as far as I'm aware.

hostep commented:
Aaaaaah ok, the github interface tricked me, I assumed it was removed when I saw all those red blocks 😄 
Okay, thanks for the explanation, those new lines in there will probably solve it (untested), sorry about the noise! 🙂

shiftedreality commented:
Sure, thank you for paying so much attention and passion for Magento!

ihor-sviziev commented:
@shiftedreality thx for good explanation!
